pkg/apis/core/validation/validation.go

```diff
@@ -2549,23 +2549,52 @@ func validateContainerPorts(ports []core.ContainerPort, fldPath *field.Path) fie
 }
 
 // ValidateEnv validates env vars
-func ValidateEnv(vars []core.EnvVar, fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
+func ValidateEnv(vars []core.EnvVar, oldPodEnvVarNames sets.Set[string], fldPath *field.Path, opts PodValidationOptions) field.ErrorList {
 	allErrs := field.ErrorList{}
 
 	for i, ev := range vars {
 		idxPath := fldPath.Index(i)
 		if len(ev.Name) == 0 {
 			allErrs = append(allErrs, field.Required(idxPath.Child("name"), ""))
 		} else {
-			for _, msg := range validation.IsEnvVarName(ev.Name) {
-				allErrs = append(allErrs, field.Invalid(idxPath.Child("name"), ev.Name, msg))
+			if utilfeature.DefaultFeatureGate.Enabled(features.RelaxedEnvironmentVariableValidation) || relaxedEnvVarUsed(ev.Name, oldPodEnvVarNames) {
```

Comment:
For completeness, check the "options" here rater than the gate.  I think this will remove the need to do a lot of the plumbing you did (I'm so sorry - I didn't know this pattern).

Specifically, look at pkg/registry/core/pod/strategy.go `podStrategy.Validate)`.  It does:

```
    opts := podutil.GetValidationOptionsFromPodSpecAndMeta(&pod.Spec, nil, &pod.ObjectMeta, nil)
    opts.ResourceIsPod = true
    return corevalidation.ValidatePodCreate(pod, opts)
```

For things like Job, we already do:
```
opts := validationOptionsForJob(job, oldJob)
```

where `JobValidationOptions` embeds `PodValidationOptions`

So aaaaaaallllll those places you fixed to pass "old" maybe don't need it.  Instead you check this in pkg/api/pod/util.go

Reply:
Okay, I'll finish it soon.